Moving SIMD types into Kokkos

ibaned opened the pull request:
This is a draft so that others may review the ongoing effort. The idea is to consolidate SIMD type work in [simd-math](https://github.com/kokkos/simd-math), [P3A](https://github.com/sandialabs/p3a), and other Sandia software and put it into the "core" Kokkos repository as a sub-package similar to "algorithms" or "containers".

The reasoning for making it part of the "core" Kokkos repository is to ease access of these tools to teams that use Kokkos through Trilinos and to ease maintenance of these tools by the Kokkos team.

crtrott commented:
We probably should have it in the Experimental namespace first?

ibaned commented:
> We probably should have it in the Experimental namespace first?

Yes, I have that in my mental todo list

ibaned commented:
I believe this is ready for review, feedback, and testing. In discussions with Christian it sounds like we would like to get this minimal starting set of functionality in now and add more in future pull requests instead of making this bigger.

masterleinad commented:
It would be a good start to get this passing CI. `jenkins` CI won't run before the code is indented with clang-format 8.0.0. :slightly_smiling_face: 
Also, changes from #5091 are still contained in this pull request. Please rebase on top of `develop`.

ibaned commented:
Is there any CI testing with C++17? I'm making this enabled only when C++17 is used

masterleinad commented:
> Is there any CI testing with C++17? I'm making this enabled only when C++17 is used

Jenkins tests CUDA, HIP, SYCL and OpenMPTarget with C++17 support.

masterleinad commented:
See https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/8950/pipeline/49/ for the latest commit for example.

dalg24 commented on CMakeLists.txt:
```diff
@@ -194,7 +194,12 @@ KOKKOS_SETUP_BUILD_ENVIRONMENT()
 
 SET(KOKKOS_EXT_LIBRARIES Kokkos::kokkos Kokkos::kokkoscore Kokkos::kokkoscontainers Kokkos::kokkosalgorithms)
-SET(KOKKOS_INT_LIBRARIES kokkos kokkoscore kokkoscontainers kokkosalgorithms)
+SET(KOKKOS_SUB_LIBRARIES kokkos kokkoscore kokkoscontainers kokkosalgorithms)
```
What is the rational for introducing `KOKKOS_SUB_LIBRARIES`?

ibaned replied:
Later on the target `kokkos` is linked to `${KOKKOS_SUB_LIBRARIES}`, as a way to conditionally exclude the `kokkossimd` target from being linked to `kokkos`.

dalg24 commented on CMakeLists.txt:
```diff
@@ -194,7 +194,12 @@ KOKKOS_SETUP_BUILD_ENVIRONMENT()
+  LIST(APPEND KOKKOS_EXT_LIBRARIES Kokkos::kokkossimd)
+  LIST(APPEND KOKKOS_SUB_LIBRARIES kokkossimd)
+ENDIF()
+SET(KOKKOS_INT_LIBRARIES kokkos ${KOKKOS_SUB_LIBRARIES})
```
`kokkos` is listed twice?  Is that intentional?

ibaned replied:
Ah shoot, no that was the whole point of it being a different variable. Will fix!

dalg24 commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,224 @@
+
+class element_aligned_tag {};
+
+template <class Mask, class Value>
+class const_where_expression {
```
Isn't that class template an implementation detail?  If so, move it to an `Impl::` namespace.

dalg24 commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,224 @@
+template <class T, class Abi>
+class simd_mask;
+
+class element_aligned_tag {};
```
Prefer
```suggestion
struct element_aligned_tag {};
```
for tag classes.

dalg24 commented on simd/src/Kokkos_Simd_Scalar.hpp:
```diff
@@ -0,0 +1,299 @@
+ public:
+  KOKKOS_FORCEINLINE_FUNCTION
+  const_where_expression(mask_type const& mask_arg, value_type const& value_arg)
+      : m_value(const_cast<value_type&>(value_arg)), m_mask(mask_arg) {}
```
Why are you casting the `const` away?  Where do you need non-`const` access?

ibaned replied:
This is copied from another effort to implement `std::simd`:
https://github.com/VcDevel/std-simd/blob/c69cb8f6a8627c186427b08662b05693176c73b2/experimental/bits/simd.h#L3096

The issue is that `where_expression` needs to derive from `const_where_expression`, so to avoid having a duplicate non-const reference to the SIMD value in `where_expression`, `const_where_expression` actually stores a non-const reference to the SIMD value and uses `const_cast`. It is not my favorite design but it seems to be what the "expert" settled on

dalg24 commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,224 @@
+}
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION const_where_expression<
```
Why "force inline"?  (Here and everywhere below)

ibaned replied:
In the P3A library that I copied this from, I found that this made a significant performance difference over just "inline" for other methods in the code (for example functor `operator()`), so I've gotten into the habit of marking anything that should ideally disappear as a compile-time abstraction with force inline. This gives me peace of mind that calls to `simd` types will decay into just intrinsic instruction calls.

dalg24 commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,224 @@
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION const_where_expression<
+    simd_mask<T, Abi>, simd<T, Abi>>
+where(desul::Impl::dont_deduce_this_parameter_t<simd_mask<T, Abi>> const& mask,
```
No, do not use implementation details from `desul`.  I am aware that they are a view occurrences in Kokkos but let's not introduce more violations.
Either define your own, or use `Impl::identity<T>` from `<impl/Kokkos_Utilities.hpp>`.  I really dislike the name but that is essentially a backport from the C++20 facility `std::type_identity`.

ibaned replied:
Okay, will switch to `Impl::identity<T>`, thanks!

dalg24 commented on simd/src/Kokkos_Simd_Scalar.hpp:
```diff
@@ -0,0 +1,299 @@
+  using abi_type                        = simd_abi::scalar;
+  using reference                       = value_type&;
+  KOKKOS_DEFAULTED_FUNCTION simd_mask() = default;
+  KOKKOS_FORCEINLINE_FUNCTION static constexpr std::size_t size() { return 1; }
```
```suggestion
  KOKKOS_FUNCTION static constexpr std::size_t size() { return 1; }
```
Same comment for all member functions defined in the body of the class.

ibaned replied:
As I commented above, I feel that FORCEINLINE is useful for these methods, but I'm willing to profile and switch to regular `KOKKOS_FUNCTION`

dalg24 commented on core/src/Kokkos_Macros.hpp:
```diff
@@ -345,9 +345,9 @@
-#if !defined(KOKKOS_IMPL_FORCEINLINE_FUNCTION)
-#define KOKKOS_IMPL_FORCEINLINE_FUNCTION inline __attribute__((always_inline))
-#define KOKKOS_IMPL_FORCEINLINE __attribute__((always_inline))
+#if !defined(KOKKOS_HOST_FORCEINLINE_FUNCTION)
```
Why did you introduce this guy?

ibaned replied:
because `KOKKOS_FORCEINLINE_FUNCTION` includes `__host__ __device__` when CUDA is enabled, and so I can't use it on the methods of the AVX-512 backend because they are host-only

crtrott replied:
Lets make that impl

dalg24 commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,224 @@
+  T a_array[simd<T, Abi>::size()];
+  a.copy_to(a_array, element_aligned_tag());
+  for (int i = 0; i < simd<T, Abi>::size(); ++i) {
+    a_array[i] = std::exp(a_array[i]);
```
Prefer `Kokkos::exp` from `<Kokkos_MathematicalFunctions.hpp>`
Same comment for all math functions below.

ibaned replied:
Agreed, will switch

dalg24 commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,224 @@
+  }
+  a.copy_from(a_array, element_aligned_tag());
+  return a;
+}
```
Do we anticipate more math functions?

ibaned replied:
Yes, I think some of what is needed by STK is still missing

ibaned replied:
Also, as a future PR I would like to reimplement them using `operator[]` instead of `copy_from`

ibaned commented:
@dalg24 we are getting the following compilation error in the `OPENMPTARGET` build of the `continous-integration/jenkin-ci/pr` check:
```
/var/jenkins/workspace/Kokkos/core/src/Kokkos_Atomics_Desul_Wrapper.hpp:84:101: error: call to 'atomic_fetch_sub' is ambiguous

T atomic_fetch_sub (T* const dest, desul::Impl::dont_deduce_this_parameter_t<const T> val) { return desul::atomic_fetch_sub (dest, val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }

                                                                                                    ^~~~~~~~~~~~~~~~~~~~~~~

/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_HostBarrier.hpp:124:13: note: in instantiation of function template specialization 'Kokkos::atomic_fetch_sub<int>' requested here

    Kokkos::atomic_fetch_sub(buffer + arrive_idx, size);

            ^

/var/jenkins/workspace/Kokkos/core/src/../../tpls/desul/include/desul/atomics/GCC.hpp:87:1: note: candidate function [with T = int]

DESUL_GCC_INTEGRAL_OP_ATOMICS(MemoryOrderRelaxed, MemoryScopeDevice)

^

/var/jenkins/workspace/Kokkos/core/src/../../tpls/desul/include/desul/atomics/GCC.hpp:30:64: note: expanded from macro 'DESUL_GCC_INTEGRAL_OP_ATOMICS'

  typename std::enable_if<std::is_integral<T>::value, T>::type atomic_fetch_sub(  \

                                                               ^

/var/jenkins/workspace/Kokkos/core/src/../../tpls/desul/include/desul/atomics/openmp/OpenMP_40_op.inc:13:56: note: candidate function [with T = int]

  std::enable_if_t<Impl::is_openmp_atomic_type_v<T>,T> atomic_fetch_sub(  

                                                       ^

7 errors generated.
```
I'm pretty sure this is unrelated to this PR. Could someone please look into this?

ibaned commented:
> @dalg24 we are getting the following compilation error in the `OPENMPTARGET` build of the `continous-integration/jenkin-ci/pr` check:
> 
> ```
> /var/jenkins/workspace/Kokkos/core/src/Kokkos_Atomics_Desul_Wrapper.hpp:84:101: error: call to 'atomic_fetch_sub' is ambiguous
> 
> T atomic_fetch_sub (T* const dest, desul::Impl::dont_deduce_this_parameter_t<const T> val) { return desul::atomic_fetch_sub (dest, val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
> 
>                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~
> 
> /var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_HostBarrier.hpp:124:13: note: in instantiation of function template specialization 'Kokkos::atomic_fetch_sub<int>' requested here
> 
>     Kokkos::atomic_fetch_sub(buffer + arrive_idx, size);
> 
>             ^
> 
> /var/jenkins/workspace/Kokkos/core/src/../../tpls/desul/include/desul/atomics/GCC.hpp:87:1: note: candidate function [with T = int]
> 
> DESUL_GCC_INTEGRAL_OP_ATOMICS(MemoryOrderRelaxed, MemoryScopeDevice)
> 
> ^
> 
> /var/jenkins/workspace/Kokkos/core/src/../../tpls/desul/include/desul/atomics/GCC.hpp:30:64: note: expanded from macro 'DESUL_GCC_INTEGRAL_OP_ATOMICS'
> 
>   typename std::enable_if<std::is_integral<T>::value, T>::type atomic_fetch_sub(  \
> 
>                                                                ^
> 
> /var/jenkins/workspace/Kokkos/core/src/../../tpls/desul/include/desul/atomics/openmp/OpenMP_40_op.inc:13:56: note: candidate function [with T = int]
> 
>   std::enable_if_t<Impl::is_openmp_atomic_type_v<T>,T> atomic_fetch_sub(  
> 
>                                                        ^
> 
> 7 errors generated.
> ```
> 
> I'm pretty sure this is unrelated to this PR. Could someone please look into this?

Nevermind, I'm pretty sure it is caused by me including Desul headers directly. working on a fix.

ibaned commented:
@dalg24 it looks like this time something interrupted the test process? This `jenkin-ci` thing seems quite fragile...

dalg24 commented:
Essentially, it now needs reviews.  I will give it another pass later this week but my approval alone does not meet our requirements to merge, we need other developers to look at it too.

masterleinad commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+template <class T>
+[[nodiscard]] KOKKOS_HOST_FORCEINLINE_FUNCTION bool any_of(
+    simd_mask<T, simd_abi::avx512_fixed_size<8>> const& a) {
+  static const __mmask16 false_value(-std::int16_t(false));
+  const __mmask16 a_value(0x0000 | static_cast<__mmask8>(a));
+  return !_kortestc_mask16_u8(~a_value, false_value);
```
Why are you using `__mmask16` here?

ibaned replied:
Good catch. I think this was done by the STK team to support Xeon Phi where AVX512DQ is not provided (apparently `kortestc_mask8_u8` is only provided with AVX512DQ), but I guess I require AVX512DQ in this implementation so I should be able to use the mask8 one

Rombur commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,262 @@
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION simd<T, Abi> exp(simd<T, Abi> a) {
+  T a_array[simd<T, Abi>::size()];
+  a.copy_to(a_array, element_aligned_tag());
+  for (int i = 0; i < simd<T, Abi>::size(); ++i) {
```
Is `size()` an `int`? Is that what the standard will do?

ibaned replied:
ooops, no. The standard says `size()` returns `std::size_t`. I can change this loop.

masterleinad commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+  KOKKOS_HOST_FORCEINLINE_FUNCTION explicit simd(
+      simd<std::uint64_t, abi_type> const& other);
+  template <class G,
+            typename std::enable_if<
+                // basically, can you do { value_type r =
+                // gen(std::integral_constant<std::size_t, i>()); }
+                std::is_invocable_r_v<value_type, G,
+                                      std::integral_constant<std::size_t, 0>>,
+                bool>::type = false>
```
```suggestion
            std::enable_if_t<
                // basically, can you do { value_type r =
                // gen(std::integral_constant<std::size_t, i>()); }
                std::is_invocable_r_v<value_type, G,
                                      std::integral_constant<std::size_t, 0>>,
                bool> = false>
```
here and elsewhere.

masterleinad commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+};
+
+KOKKOS_HOST_FORCEINLINE_FUNCTION
+simd<std::uint64_t, simd_abi::avx512_fixed_size<8>> condition(
```
Is this standardized? Otherwise,
```suggestion
simd<std::uint64_t, simd_abi::avx512_fixed_size<8>> blend(
```
and some comments what this does would be good. :slightly_smiling_face:

ibaned replied:
This is not standardized. The latest conversation related to standardization:
https://github.com/mattkretz/std-simd-feedback/issues/17
Says that they would rely on being able to overload `operator?:`.

This function is used a lot in my physics application and in others at Sandia, so a wider conversation would be needed to rename it. I chose `condition` because `operator?:` is sometimes called the conditional operator. Similarly, `std::conditional` chooses one of two types based on a condition.

masterleinad replied:
If this is already used a lot, that's fine with me.
I just found it quite difficult to understand what this is doing without looking up the intrinsic you are using and without any comments.

ibaned replied:
Good point, I can add a comment on overloads and make sure to prominently document it in the future PR to the Kokkos Wiki

masterleinad commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+}
+
+template <>
+class const_where_expression<simd_mask<double, simd_abi::avx512_fixed_size<8>>,
+                             simd<double, simd_abi::avx512_fixed_size<8>>> {
```
https://en.cppreference.com/w/cpp/experimental/simd/const_where_expression also has `operator-` here.

ibaned replied:
Yes, I would be happy to add them if demand is indicated. This pretty much just boils down to limited time.

masterleinad commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+};
+
+template <>
+class where_expression<simd_mask<double, simd_abi::avx512_fixed_size<8>>,
+                       simd<double, simd_abi::avx512_fixed_size<8>>>
```
https://en.cppreference.com/w/cpp/experimental/simd/where_expression has a bunch more member functions. Are you intending to add them eventually?

ibaned replied:
Would be happy to aim for completeness in `where_expression` operators, just did not due to limited time

masterleinad commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,262 @@
+  return a;
+}
+
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION constexpr bool all_of(bool a) {
+  return a;
+}
+
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION constexpr bool any_of(bool a) {
+  return a;
+}
```
What do you need these for?

ibaned replied:
This is actually really important (and specified in the ISO TS) so that I can write code like this:
```cpp
template <class T>
void do_expensive_physics_if(T should_do)
{
  if (any_of(should_do)) {
    // if any of the lanes are true then do lots of vectorized math
    // but if none of the lanes are true then we can skip this work
  }
}
```
And it works when `T=bool` and when `T=simd_mask<...>`, basically generic code that works with SIMD types or regular arithmetic types

masterleinad commented on simd/src/Kokkos_Simd_Scalar.hpp:
```diff
@@ -0,0 +1,353 @@
+                                     : identity_element;
+}
+
+// these are specialized because NumericTraits is in KokkosKernels not Kokkos
```
What do you mean?

ibaned replied:
Ideally these would use `std::numeric_traits<T>::max()` and friends. On CUDA this isn't possible unless you enable experimental `constexpr` support in CUDA. The comment refers to my desire to have `Kokkos::numeric_traits` in `kokkoscore` and use it here.

ibaned commented on simd/src/Kokkos_Simd_Common.hpp:
```diff
@@ -0,0 +1,262 @@
+  return a;
+}
+
+template <class U, class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION
+    std::enable_if_t<std::is_arithmetic_v<U>, simd<T, Abi>>
+    operator+(U const& a, simd<T, Abi> const& b) {
+  return simd<T, Abi>(T(a)) + b;
+}
```
I think this was to emulate arithmetic promotions since the compiler doesn't twice-implicit-cast on its own. For example:
```cpp
auto minus_42 = -1 * simd<double,native>(42.0);
```
Where the left hand side is type `int`... I'm not 100% sure anymore though.

ibaned replied:
Yep, I just removed this to confirm, and I could no longer add `int` to a `simd<double,...>`.

ibaned replied:
Your original comment was correct, I made it work the way it is supposed to now

ibaned commented on simd/cmake/Dependencies.cmake:
```diff
@@ -0,0 +1,5 @@
+TRIBITS_PACKAGE_DEFINE_DEPENDENCIES(
+  LIB_REQUIRED_PACKAGES KokkosCore
+  LIB_OPTIONAL_TPLS Pthread CUDA HWLOC HPX
```
These are Trilinos TPLs, I don't know if `HIP` and `SYCL` are TPLs in Trilinos. Even if this is true though, I think it is outside the scope of this PR.

ibaned commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+
+template <>
+class simd<std::uint32_t, simd_abi::avx512_fixed_size<8>> {
+  __m256i m_value;
```
Because it is `avx512_fixed_size<8>`, so there must be 8 of the `value_type`, so it is a 256-bit register to store 8 32-bit integers. This is the common case of folks using 8 32-bit integers to do scatter or gather operations on 8 64-bit doubles.

ibaned commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+    return reinterpret_cast<value_type const*>(&m_value)[i];
+  }
+  KOKKOS_HOST_FORCEINLINE_FUNCTION simd operator*(simd const& other) const {
+    return simd(_mm256_mullo_epi32(m_value, other.m_value));
```
See the comment above about "Why not __m512i?". Basically, the ABI type in this file is all about dealing with 8 things, and if those things are 32-bit then that means using 256-bit vector instructions

ibaned commented on simd/src/Kokkos_Simd.hpp:
```diff
@@ -0,0 +1,96 @@
+using host_native   = scalar;
+#endif
+
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
```
Someone else asked this, yes, I think a more general macro can apply here

ibaned commented on simd/src/Kokkos_Simd_AVX512.hpp:
```diff
@@ -0,0 +1,928 @@
+
+#include <immintrin.h>
+
+#ifndef __AVX512DQ__
+#error "Kokkos AVX512 SIMD backend requires AVX512DQ"
+#endif
```
I think Skylake is supposed to support AVX512DQ, so this should be impossible unless people are saying their CPU is Skylake or later and it is not

masterleinad replied:
Yeah, I would expect `Skylake`, `Ice Lake`, and `Sapphire Rapids` all to support `AVX512DQ`.

ibaned commented:
Thank you all for the second round of review, I will begin making commits and comment when done.

ibaned commented:
Alright, I believe I've made all the changes I intended to make in response to the second round of review.

ibaned commented:
In my opinion this is ready to merge, with feedback from the Kokkos team addressed as much as possible. All CI is actually passing, the SYCL OneAPI test just timed out again at 6 hours...

ibaned commented:
Some TODO notes from a meeting:
- [ ] fix leftover `KOKKOS_HOST_FORCEINLINE_FUNCTION`
- [ ] remove `host_native`, `device_native`
- [ ] add some kind of native trait given `ExecutionSpace`
- [ ] typedef `native` to the trait for `DefaultExecutionSpace`
- [ ] move `cmath` overloads to `Kokkos::`
- [ ] check on the implementation of `cmath` overloads versus STK

vbrunini commented on simd/unit_tests/TestSIMD.cpp:
```diff
@@ -0,0 +1,376 @@
+TEST(simd, device) {
+  Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::IndexType<int>>(0, 1),
+                       simd_device_functor());
+}
```
I only skimmed the implementation files, but I think they largely look fine.

I've looked more at the tests since those are the best way to ensure we have all the desired functionality as we add backends etc. Here's what I see missing right now:
1) The compound assignment operators (+=,-=, *=,/=)
2) Comparison operators
3) All the math functions (max/min/abs/acos/asin/atan/cos/exp/log/pow/sin/sqrt/tan/tanh are the ones I know we use).
4) All of the above where one or more of the operands are non-SIMD types.

I don't feel strongly about including them in this PR or as follow-up.

ibaned replied:
Thanks Victor! Most of that is there but not covered in tests, but what is not there is mainly the full `<cmath>` functions. Maybe given the slowness of Kokkos PR review I will add them.

masterleinad commented:
Looks like the indentation is off. :slightly_smiling_face:

tasmith4 reviewed (commented):
@ibaned nothing stands out to me as an obvious issue, but I'll let someone from the Kokkos team approve for real

masterleinad reviewed (changes requested):
Fix the indentation.

ibaned commented:
Alright, once again this passes all CI and has addressed all the feedback given so far including math functions, indentation, etc. In my opinion we could merge this now.

Rombur reviewed (approved):
Looks good

ibaned commented:
Alright, I think I've figured out how to "disable" `KokkosSimd` when C++17 is not available in the TriBITS/Trilinos build. The latest changes allow to build with and without C++17 in Trilinos.

ibaned commented:
This is once again ready to merge in my opinion, including Trilinos support. the only failure in the Jenkins CI is due to the test `KokkosCore_UnitTest_CudaTimingBased` under NVHPC, which I believe must be unrelated.

dalg24 commented on cmake/Dependencies.cmake:
```diff
@@ -1,3 +1,14 @@
+  SET(KOKKOS_SIMD_TEST_CLASS PT)
+ELSE()
+  SET(KOKKOS_SIMD_TEST_CLASS EX)
+  IF (${PROJECT_NAME}_ENABLE_KokkosSimd)
```
Any reason why you preferred `${PROJECT_NAME}_` to `Kokkos_`?

ibaned replied:
Because this is specifically for Trilinos and in Trilinos `${PROJECT_NAME}=Trilinos`, while in the standalone build `${PROJECT_NAME}=Kokkos`.

dalg24 replied:
I am not sure it is a good idea though, because it is at the end of the day a Kokkos option and in a world where Kokkos becomes a TPL for Trilinos it will surprise users.
Trilinos packages are free to introduce their own dependent option after the fact if they like to do so.

ibaned replied:
It is not just a good idea, it is the only correct solution. We are not in a world where Kokkos is a TPL for Trilinos. Kokkos is a Trilinos package that gets snapshotted into Trilinos. When that happens this file decides whether the `KokkosSimd` sub-package of Trilinos is disabled by default or not, and Trilinos users *have to* use `Trilinos_ENABLE_KokkosSimd` to enable it otherwise, `Kokkos_ENABLE_KokkosSimd` does nothing. I am not introducing anything new or confusing here, I am working with the TriBITS build that Kokkos has to support for now.

ibaned replied:
This is not a Kokkos option. It is only an option in the TriBITS/Trilinos build.
